### PR TITLE
GODRIVER-1859 Add ignore writeErrors SDAM spec test

### PR DESCRIPTION
**Pull request by benjirewis**
[GODRIVER-1859](https://jira.mongodb.org/browse/GODRIVER-1859)

Adds a spec test to ensure the writeErrors field is ignored in server discovery and monitoring.

It looks like the Go driver already has the correct behavior.

### CODE REVIEW

**Review by iwysiu — APPROVED** [review]

**Review by divjotarora — COMMENTED** [review]
The SDAM spec tests use a [getError](https://github.com/mongodb/mongo-go-driver/blob/ea02175e9e44795efdda597ebd94b11d746d3984/x/mongo/driver/topology/sdam_spec_test.go#L307) helper function to extract an error from a BSON document that represents a server response. This code is basically copied from the `driver` package [extractError](https://github.com/mongodb/mongo-go-driver/blob/ea02175e9e44795efdda597ebd94b11d746d3984/x/mongo/driver/errors.go#L336) function because it's not exported. However, `getError` ignores any errors if the `ok` value in the response is 1 (see [here](https://github.com/mongodb/mongo-go-driver/blob/ea02175e9e44795efdda597ebd94b11d746d3984/x/mongo/driver/topology/sdam_spec_test.go#L323)) so in the case of the new test you added, it's returning `nil`, which is a no-op and isn't actually testing our processing of `writeErrors`.

My recommendation would be to export the `extractError` function with a new name that clarifies its purpose and use it here directly. This would also take care of [GODRIVER-1784](https://jira.mongodb.org/browse/GODRIVER-1784) for us at the same time.

**Comment by benjirewis** [comment]
Sounds good, thanks for the explanation. I've exported `extractError` as `ExtractErrorFromBsoncoreDoc` and removed the original `getErrors` helper function. It should actually be testing that we don't process `writeErrors` for SDAM now...

File: x/mongo/driver/errors.go [hunk 1]
```diff
@@ -331,16 +331,16 @@ func (e Error) NamespaceNotFound() bool {
 	return e.Code == 26 || e.Message == "ns not found"
 }
 
-// helper method to extract an error from a reader if there is one; first returned item is the
-// error if it exists, the second holds parsing errors
-func extractError(rdr bsoncore.Document) error {
+// ExtractErrorFromBsoncoreDoc extracts an error from a bsoncore.Document if there is one.
+// Also used in testing for SDAM.
+func ExtractErrorFromBsoncoreDoc(doc bsoncore.Document) error {
```
Comment (divjotarora) [inline x/mongo/driver/errors.go:336]: We want functions names to describe their purpose in as much detail as possible. This function name is technically correct, but it doesn't provide details about what the bsoncore document should actually contain. Also, the function API already takes `bsoncore.Document`, so `BsoncoreDoc` in the function name doesn't provide any new information. Instead, the name should describe the document's semantics and invariants. I'd suggest renaming to something like `ExtractErrorFromServerResponse`. Thoughts?

Comment (benjirewis) [reply x/mongo/driver/errors.go:336]: `ExtractErrorFromServerResponse` seems appropriate to me; all three current usages seem to do exactly that, and it's true that `BsoncoreDoc` is redundant.

**Review by divjotarora — APPROVED** [review]

File: x/mongo/driver/topology/sdam_spec_test.go [hunk 2]
```diff
@@ -304,147 +304,12 @@ func (n netErr) Temporary() bool {
...
-	}
-}
-
 func applyErrors(t *testing.T, topo *Topology, errors []applicationError) {
 	for _, appErr := range errors {
 		var currError error
 		switch appErr.Type {
 		case "command":
-			currError = getError(appErr.Response)
+			currError = driver.ExtractErrorFromServerResponse(appErr.Response)
```
Comment (kevinAlbs) [inline x/mongo/driver/topology/sdam_spec_test.go:312]: Ah nice, using the newly exported `ExtractErrorFromServerResponse` seems so much better than using `getError` copy.

**Review by kevinAlbs — APPROVED** [review]
LGTM!

**Pull request merged**